### PR TITLE
add style object basics from music21

### PR DESCRIPTION
**Pull request by mscuthbert**
lets .style.hideObjectOnPrint work on clefs, ts, ks.

Leaves extra space on the Stream however for the hidden objects.  A to-do for later

Screen Shot 2021-12-18 at 07 59 03

### CODE REVIEW

File: src/base.ts [hunk 1]
```diff
@@ -165,10 +168,46 @@ export class Music21Object extends prebase.ProtoM21Object {
...
+     * Returns true if there is a style.Style object
+     * already associated with this object, false otherwise.
+     *
+     * Calling .style on an object will always create a new
+     * Style object, so even though a new Style object isn't too expensive
+     * to create, this property helps to prevent creating new Styles more than
+     * necessary.
+     */
+    get hasStyleInformation(): boolean {
+        return !(this._style === undefined);
```
Comment (jacobtylerwalls) [inline src/base.ts:185]: I know it's picky to ask for it, but how about `!==`?

File: src/stream.ts [hunk 2]
```diff
@@ -2028,13 +2028,16 @@ export class Stream extends base.Music21Object {
         }
         const rendOp = this.renderOptions;
         if (rendOp.displayClef) {
+            // TODO: do not do this if Clef.style.hideObjectOnPrint
             totalLength += 30;
         }
         if (rendOp.displayKeySignature) {
+            // TODO: do not do this if Clef.style.hideObjectOnPrint
```
Comment (jacobtylerwalls) [inline src/stream.ts:2035]: remove or replace Clef with KeySignature

File: src/stream.ts [hunk 3]
```diff
@@ -2028,13 +2028,16 @@ export class Stream extends base.Music21Object {
...
             const ks = this.getSpecialContext('keySignature') || this.getContextByClass('KeySignature');
             totalLength += ks?.width ?? 0;
         }
         if (rendOp.displayTimeSignature) {
+            // TODO: do not do this if Clef.style.hideObjectOnPrint
```
Comment (jacobtylerwalls) [inline src/stream.ts:2040]: sim.

File: src/style.ts [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+// noinspection JSUnusedGlobalSymbols
+export class Style extends prebase.ProtoM21Object {
+    size: undefined;
+    relativeX: number|undefined;
+    relativeY: number|undefined;
+    absoluteX: number|undefined;
+    absoluteY: number|undefined;
+    enclosure: Enclosure|undefined;
+    // fontRepresentation
+    color: string|undefined;
```
Comment (jacobtylerwalls) [inline src/style.ts:31]: No issue with this. Just asking for when it comes up in future: should we be adding `|undefined` like this in the event we turn on ``--strict`` in the ts compiler? Right now with strict=false, undefined is always implied. (And we're a long way from turning strict mode on.)

File: src/style.ts [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    OCTAGON = 'octagon',
+    NONAGON = 'nonagon',
+    DECAGON = 'decagon',
+    NONE = 'none',  // special.  sets to undefined
+}
+
+
+// noinspection JSUnusedGlobalSymbols
+export class Style extends prebase.ProtoM21Object {
+    size: undefined;
```
Comment (jacobtylerwalls) [inline src/style.ts:24]: this one seems unnecessary to type. also, I looked in m21p and we never set or get it, so maybe we just jettison?

File: src/stream.ts (hunk 2 above)
Comment (jacobtylerwalls) [reply src/stream.ts:2035]: but also, not sure that we want to change this, so maybe QUESTION

**Review by jacobtylerwalls — APPROVED** [review]
Super!

Some nitpicks you can feel free to ignore.

> Leaves extra space on the Stream however for the hidden objects. A to-do for later

I'm not sure we need to bother. Setting `.renderOptions.displayTimeSignature = false` should take care of the spacing issue (as well as display, without having to resort to `hideObjectOnPrint` AFAIK). If someone wants to hide something, I think intuition could cut either way re: whether to alter the spacing--someone might want to just drop in their own element at that location.

File: src/style.ts (hunk 4 above)
Comment (mscuthbert) [reply src/style.ts:31]: I want to do this and eventually turn on `--strict`.  We're a long way from turning it on, but I want to show places where undefined is a totally acceptable value and not just a "we haven't gotten around to defining it yet"

File: src/style.ts (hunk 5 above)
Comment (mscuthbert) [reply src/style.ts:24]: Oops, forgot to define a value for it.  Will comment out for now, but want to remember that it should be there eventually.

Btw -- I've also rewritten Enclosure as a StrEnum in Python, but I think I need to make a v8 branch before committing it.

**Comment by mscuthbert** [comment]
Right now, the issue is that renderOptions.displayClef gets overwritten every time the Stream's position is recalculated, so until we separate out renderOptions from style, we can't just rely on users setting renderOptions.

**Comment by jacobtylerwalls** [comment]
> renderOptions.displayClef gets overwritten every time the Stream's position is recalculated

Ah right, okay, sounds like #135. Got it.

**Pull request merged**